### ISSUE TITLE
Rename and publicize `Base.Filesystem._readdirx`

### ISSUE THREAD

**Issue by LilithHafner** [JuliaLang/julia#55333]
[We added _readdirx](https://github.com/JuliaLang/julia/pull/53377#issuecomment-1970289565) with an internal name to defer bikeshedding. Here's an issue for that bikeshed.

**Comment by LilithHafner** [comment]
Triage talked about this and came up with the following:

- `readdir(DirEntry, path)`

  - More discoverable on `?readdir`
  - Extensible, e.g. `readdir(Path, path)` for a package that defines paths
  - Julia is all about shoving things into functions
  - Follows `parse` and `JuliaSyntax.parse`

- `readdirentry(path)`

  - More clear that it is different functionality

---

Nobody at triage wanted these names:

- `readdirx(path)`
- `ls(path)`
- `parse(Vector{DirEntry}, path)`

**Comment by ericphanson** [comment]
it would be nice if there was an abstraction also usable by S3, i.e. some `readdir`/`walkdir` style function that when used with s3 can give you back an iterable of objects which knows the metadata that we got for-free from `list_objects_v2()` or such (i.e. key, size, last modified, etag, etc).

I think `readdir(DirEntry, path)` could work for that where AWSS3 could define `readdir(S3Entry, path::S3Path)`.

**Comment by oscardssmith** [comment]
I think `readdir(S3Entry, path::S3Path)` is a pretty great example of why `readdir(DirEntry, path)` is the way to go.

**Comment by vtjnash** [comment]
I feel like that is actually an argument against that API. The idea that I need to write `readdir(typeof(path), path)` seems quite strange to me.

**Comment by IanButterworth** [comment]
`typeof(path), path` seems like the wrong example. You don't want `typeof(path)` you want an "Entry" type, so `readdir(S3Entry, path::S3Path)` makes sense to me, assuming that `readdir(path::S3Path)` returns a list of Strings already.

**Comment by ericphanson** [comment]
`readdir(path::S3Path)` indeed returns a list of strings, while `readdir(path::S3Path; join=true)` returns a list of `S3Path`s.

**Comment by LilithHafner** [comment]
`readdir(path::S3Path; join::Bool)` is type unstable. It could be made type stable if it used two a type argument to determine return type rather than the value of `join`.

**Comment by ericphanson** [comment]
yes it is type unstable, but that seems necessary for `join=true` to have the semantic that you can operate on the outputs of `readdir(path; join=true)` with e.g. `isfile`, `read`, etc, which is the useful property of `join`, and the network call will likely be in ms compared to the microseconds of dynamic dispatch anyway. One could also hope that const-prop could eliminate the instability.

**Comment by IanButterworth** [comment]
Realized here https://github.com/JuliaLang/julia/pull/55358
